zst_pointer_async/README.md

```diff
@@ -0,0 +1,357 @@
+# Asynchronous Components using Zero Sized Type Pointers
+
+Original author: Johnathan Van Why
+
+This document introduces a design (the "ZST-pointer" design) for asynchronous
+APIs in `libtock-rs`. It then gives an analysis of the design's tradeoffs, and
+the results of an experiment evaluating how this API design can live alongside
+a futures-based asynchronous API.
+
+## Background
+
+`libtock-rs` should expose asynchronous interfaces to the Tock kernel’s
+asynchronous APIs. There are a variety of mechanisms a library can use to expose
+an asynchronous API, with varying tradeoffs. The Rust community has settled on
+using Rust’s [Future](https://doc.rust-lang.org/core/future/trait.Future.html)
+trait to construct asynchronous APIs. Unfortunately, the analysis at [Futures
+versus no-Futures Size Comparison](../size_comparison) shows that the Future
+trait has size costs that are too high for many use cases of `libtock-rs`. As a
```

Comment:
I would not agree to that extend. The point there is _a_ implementation of _a_ future based api which produces larger binaries at small scale. However, this implementation doesn't relate to the current future based api of `libtock-rs`.

Reply:
What exactly are you looking for? I expect that the costs are too high for most of Google's use cases, so that alone probably counts as "many" (at least as a fraction of all `libtock-rs` use cases).

I don't see how futures can be implemented in a more lightweight manney than they were in that API. If we don't store a `Waker` (`libtock-rs`'s current design), we lose the ability to use existing futures (e.g. `futures::stream::FuturesUnordered`).